lib/utils/uri.go

```diff
@@ -39,3 +40,17 @@ func ParseSessionsURI(in string) (*url.URL, error) {
 	}
 	return u, nil
 }
+
+// validSubdomain is a regular expression used to validate the DNS subdomain.
+// DNS subdomains must be between 1-63 characters, must start and end with an
+// alphanumeric character, and any part can contain hyphens except the first
+// or last character.
+//
+// See the following link for more details:
+//   https://stackoverflow.com/questions/7930751/regexp-for-subdomain
+var validSubdomain = regexp.MustCompile(`^[A-Za-z0-9](?:[A-Za-z0-9\-]{0,61}[A-Za-z0-9])?$`)
+
+// ValidSubdomain validates a DNS subdomain.
+func ValidSubdomain(s string) bool {
+	return validSubdomain.MatchString(s)
+}
```

Comment:
Maybe we should also check that the length of `appname.proxy.example.com` (i.e. the app name concatenated with the proxy's public address) is not >255 characters?

Reply:
Can you create a ticket for that? It gets a little complex because it's not always clear when registering an application what full domain it will be available at. I'm specifically thinking about trusted clusters here.

Reply:
Here is some good example in k8s validation package 

https://github.com/kubernetes/apimachinery/blob/master/pkg/util/validation/validation_test.go#L91

in cloud we use `validation.IsDNS1123Subdomain` to verify the full FQDN and `IsDNS1035Label` to check a sub-domain name.

Reply:
Alright I've switched to `IsDNS1035Label` so Teleport and Cloud are in sync.